### PR TITLE
feat(fleet): add /rollout endpoint to server

### PR DESCRIPTION
**Pull request by TBonnin**
adding /rollout endpoint 

# How to test
1. make sure you have `NANGO_INTERNAL_API_KEY` env var setup locally
2. run Nango locally
3. `curl -X POST http://localhost:3003/internal/fleet/nango_runners/rollout -H 'Content-Type: application/json' -d '{ "commitHash": "7ed44fc18fc76c66631ac3d4169107b1a48f9c58" }' -H 'Authorization: Bearer XXX'
4. observer new runner to be created and old one to be terminated

### CODE REVIEW

File: packages/runner/lib/app.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ try {
     });
 
     const close = () => {
-        logger.info('Closing...');
+        logger.info(`${id} Closing...`);
```
Comment (TBonnin) [inline packages/runner/lib/app.ts:18]: unrelated to this PR but it is nice when running Nango locally to understand which runner is being terminated

File: packages/server/lib/middleware/access.middleware.ts [hunk 2]
```diff
@@ -392,6 +392,27 @@ export class AccessMiddleware {
 
         next();
     }
+
+    internal(req: Request, res: Response, next: NextFunction) {
```
Comment (TBonnin) [inline packages/server/lib/middleware/access.middleware.ts:397]: I am not sure about the `internal` name and API endpoint prefix below. Anybody has a better idea?

File: packages/fleet/lib/fleet.integration.test.ts [hunk 3]
```diff
@@ -21,7 +21,7 @@ describe('fleet', () => {
     describe('deploy', () => {
```
Comment (nalanj) [inline packages/fleet/lib/fleet.integration.test.ts]: Should that be `rollout`?

File: packages/server/lib/middleware/access.middleware.ts (hunk 2 above)
Comment (nalanj) [reply packages/server/lib/middleware/access.middleware.ts:397]: I've always used `internal` or `private`. I think I prefer `internal` though.

**Review by nalanj — APPROVED** [review]
Looks good.

**Comment by nalanj** [comment]
Oh, one other note. Should `NANGO_INTERNAL_API_KEY` be added to `.env.example`?

File: packages/server/lib/middleware/access.middleware.ts (hunk 2 above)
Comment (bodinsamuel) [reply packages/server/lib/middleware/access.middleware.ts:397]: Looks good to me:
public = public
private = dashboard
internal = only for us

File: packages/server/lib/middleware/access.middleware.ts [hunk 4]
```diff
@@ -392,6 +392,27 @@ export class AccessMiddleware {
...
+        }
+
+        const authorizationHeader = req.get('authorization');
+
+        if (!authorizationHeader) {
+            return errorManager.errRes(res, 'missing_auth_header');
+        }
+
+        const receivedKey = authorizationHeader.split('Bearer ').pop();
+        if (receivedKey !== key) {
```
Comment (bodinsamuel) [inline packages/server/lib/middleware/access.middleware.ts]: if we are paranoid `crypto.timingSafeEqual` (we haven't done it for the other methods but just thought about it)

File: packages/runner/lib/app.ts (hunk 1 above)
Comment (bodinsamuel) [reply packages/runner/lib/app.ts:18]: might be cleaner to create a child (but not sure if we handle msg concat)

File: packages/fleet/package.json [hunk 5]
```diff
@@ -18,6 +18,7 @@
     },
     "dependencies": {
         "@nangohq/utils": "file:../utils",
+        "@nangohq/types": "file:../types",
```
Comment (bodinsamuel) [inline packages/fleet/package.json]: missing tsconfig

File: packages/server/lib/controllers/fleet/postRollout.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    const params = paramsValidation.safeParse(req.params);
+    if (!params.success) {
+        res.status(400).send({ error: { code: 'invalid_uri_params', errors: zodErrorToHTTP(params.error) } });
+        return;
+    }
+
+    const { fleetId } = params.data;
+    if (fleetId === runnersFleet.fleetId) {
```
Comment (bodinsamuel) [inline packages/server/lib/controllers/fleet/postRollout.ts]: why not use the opposite early return to follow to rest of the endpoint pattern?

File: packages/server/lib/controllers/fleet/postRollout.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        return;
+    }
+
+    const params = paramsValidation.safeParse(req.params);
+    if (!params.success) {
+        res.status(400).send({ error: { code: 'invalid_uri_params', errors: zodErrorToHTTP(params.error) } });
+        return;
+    }
+
+    const { fleetId } = params.data;
```
Comment (bodinsamuel) [inline packages/server/lib/controllers/fleet/postRollout.ts]: body and params are never reassigned to the type so you can't be sure it's synchronized with the zod type

File: packages/server/lib/controllers/fleet/postRollout.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    if (fleetId === runnersFleet.fleetId) {
+        const rollout = await runnersFleet.rollout(body.data.commitHash);
+        if (rollout.isErr()) {
+            res.status(500).send({ error: { code: 'rollout_failed', message: rollout.error.message } });
+        } else {
+            res.status(200).send(rollout.value);
+        }
+        return;
+    }
+    res.status(400).send({ error: { code: 'unknown_fleet', message: 'Unknown fleet' } });
```
Comment (bodinsamuel) [inline packages/server/lib/controllers/fleet/postRollout.ts]: nitpick: to me it's either a standard `invalid_uri_params` or a 404

File: packages/server/lib/middleware/access.middleware.ts [hunk 9]
```diff
@@ -392,6 +392,27 @@ export class AccessMiddleware {
 
         next();
     }
+
+    internal(req: Request, res: Response, next: NextFunction) {
+        const key = process.env['NANGO_INTERNAL_API_KEY'];
```
Comment (bodinsamuel) [inline packages/server/lib/middleware/access.middleware.ts]: It would be nice to add to id to env parse.ts

**Review by bodinsamuel — APPROVED** [review]

File: packages/runner/lib/app.ts (hunk 1 above)
Comment (TBonnin) [reply packages/runner/lib/app.ts:18]: what do you mean by `create a child`?

File: packages/server/lib/middleware/access.middleware.ts (hunk 4 above)
Comment (TBonnin) [reply packages/server/lib/middleware/access.middleware.ts]: good idea

File: packages/server/lib/middleware/access.middleware.ts (hunk 9 above)
Comment (TBonnin) [reply packages/server/lib/middleware/access.middleware.ts]: we haven't used `parse.ts` in server so far. I've created a `packages/server/lib/env.ts` and added `NANGO_INTERNAL_API_KEY` to `parse.ts`

**Pull request merged**